README.md
# RigMon #


### What is RigMon? ###

RigMon is a quick way to monitor your mining rigs using Telegram bot.

It currently supports:
* Claymore miners (Eth, Equihash, CryptoNight)
* EWBF miner

You can also check balance and transaction history in:
* Bittrex
* Poloniex

## How do I get set up? ###

Download and install Node.js

https://nodejs.org/en/download/

Clone or download this repository

### Configuring config.json ###

Open `config.json` with your favorite text editor

Specify your currency and symbol here:

```
"currency": "usd"
"currency_symbol": "$"
```

You can use any currency you see on https://coinmarketcap.com, you have to use lowercase letters.

Specify your power cost per kW (for example, 0.09 is 9 cents for 1 kilowatthour)

`"electricity": 0.09`

Replace TELEGRAMBOT_TOKEN with your Telegram Bot token here:

`"token": "TELEGRAMBOT_TOKEN"`

### How to get Telegram Token

Talk to the BotFather https://telegram.me/botfather

Write /newbot and follow the instructions

You should get your token like this:
```
Use this token to access the HTTP API:
488814350:AAFxmfas0zOKSmDaAgAierd90-v8h_LKeF8
```
### Configuring your mining rigs

Sample configuration:

```
"clay": [
        {
            "name": "rig1",
            "host": "192.168.1.2",
            "port": 3333,
            "comments": "loginName/password",
            "power": 510,
            "coin": "ETH",
            "algo": "eth"
        }
]
```

Use `"clay":` for rigs using Claymore miner

Use `"ewbf":` for rigs using EWBF miner

Change host and port address to what you set up as your miners API.
```
"host": "127.0.0.1"
"port": 3333
```
Default values for Claymore is port 3333, EWBF is port 42000.

Use IP 127.0.0.1 if you are mining on this computer.

You can reference your rig with any comments you like, I put my remote login and password here.

`"comments": "comment"`

Put your power from the wall here in watts, for profit calculation:

`"power": 820`

Put the coin shortname and algorithm you are mining with this rig here:

`"coin": "ETH"`

`"algo": "eth"`

Currently available algos are: 

```
eth - Ethereum
eq - Equihash
cn - CryptoNight
```

### First time run

Don't forget to install Node.js

Open up your command prompt or terminal and navigate to your directory

(Easy way to open Command Prompt is to shift+right click inside the folder)

Navigate to your directory using 

```
cd *your directory*
for example:
cd C:/Users/User/rigmon
```

Then, run these commands

```
npm install
npm start
```

npm install will install required dependancies for this script and npm start will run it.

After running npm start for the first time, you have to send a direct message to the bot:

`/id`

You will receive your chat id which you will have to put to the config file here:

`"chatid": 123456`

It will enable monitoring of your rigs, in case it gets offline or stuck, you will receive a notification.

Each time you make a change to the config file, you will have to restart the bot.

Use ctrl+c to stop the running script and start it again.

You can also add your bot to a group chat and get an /id there.

### Exchange API keys

Currently supported exchanges:

Poloniex

Bittrex

Use the read-only API key without any access to trading.

This program uses the keys only to get your balance and withdrawal/deposit history.

Put your API keys here:

```
    "poloniex": {
        "key": "YOUR_API_KEY_POLONIEX",
        "secret": "YOUR_SECRET_KEY_POLONIEX"
    },
    "bittrex": {
        "key" : "YOUR_API_KEY_BITTREX",
        "secret" : "YOUR_SECRET_KEY_BITTREX"
    }
```

## Available commands
You can also use the following command list to change the list of available commands for your bot. Chat to the BotFather and use `/setcommands` and paste the text in below. All commands do need to begin with a forward slash e.g. `/status`

```
balances - Returns balances for your account.
detailed - Displays more information, like temperature and power usage (EWBF miners only).
help - Displays available commands.
history - Returns withdrawal and deposit history for the last week.
ip - Displays the current IP address of all rigs.
pools - Displays current pool servers.
profit - Displays current profit taken from whattomine.com, including electricity costs.
status - Displays the current status of all miners.
```


CHANGELOG.md
# Change Log

## 0.1 - Unreleased

### Enhancements

#### Help

* `/help` formatting has been updated. Links are now clickable/tappable if commands are registered with the BotFather.
* Updated the Available commands section and provided a BotFather command list template in the README file.

#### Balances
* Return "No Bittrex or Poloniex accounts configured." when both Bittrex and Poloniex accounts have not been configured.

#### History
* Return "No Bittrex or Poloniex accounts configured." when both Bittrex and Poloniex accounts have not been configured.

#### Statistics
* Removed pre tags for response.
* Command header is now bold.
* Statistics are now layed out over multiple lines for mining rigs.
* If Rigmon has just started and is still fetching data from the remote API and a user requests statistics, we now return "Mining rigs are not present as yet." instead of just a **Statistics** command header.

#### Detailed
* Removed pre tags for response.
* Command header is now bold.
* Offline miners will now show &#9888; e.g. "Miner X is offline &#9888;"
* If a miners graphics card has stalled (0 hash rate) we now show &#9888; for the graphics card currently not working.
* Miner graphic cards are now numbered and show the current cryptocurrency being mined e.g. "5 (ETH):"
* If Rigmon has just started and is still fetching data from the remote API and a user requests statistics, we now return "Mining rigs are not present as yet." instead of just a **Detailed Statistics** command header.

### Security

* Bump up version for moment from 2.19.1 to fix ReDOS vulnerability. See [Moment #4163](https://github.com/moment/moment/issues/4163)


ewbf.js
var config = require('./config.json')
var net = require('net')

var ewbf = []
ewbf.json = []

function checkJSON (jsonString) {
  try {
    var o = JSON.parse(jsonString)
    if (o && typeof o === 'object') {
      return true
    }
  } catch (e) { }

  return false
};

config.ewbf.forEach(function (item, i, arr) {
  // settings
  var nvm = ewbf[i] = {}
  var c = config.ewbf[i]
  var j = ewbf.json[i]

  nvm.name = c.name
  nvm.host = c.host
  nvm.port = c.port
  nvm.poll = (typeof c.poll !== 'undefined') ? c.poll : config.miner_poll
  nvm.timeout = (typeof c.timeout !== 'undefined') ? c.timeout : config.miner_timeout

  function hostname () {
    return c.hostname ? c.hostname : (nvm.host + ':' + nvm.port)
  }

  // stats
  nvm.reqCnt = 0
  nvm.rspCnt = 0

  // it was never seen and never found good yet
  c.last_seen = null
  c.last_good = null

  // socket
  nvm.socket = new net.Socket()

    .on('connect', function () {
      var req = '{"id":0, "method":"getstat"}'
      ++nvm.reqCnt
      nvm.socket.write(req + '\n')
      nvm.socket.setTimeout(nvm.timeout)
    })

    .on('timeout', function () {
      nvm.socket.destroy()
      ewbf.json[i] = {
        'name': nvm.name,
        'host': hostname(),
        'uptime': '',
        'pools': '',
        'ver': '',
        'comments': c.comments,
        'offline': c.offline,
        'warning': null,
        'error': 'Error: no response',
        'power': '',
        'gpu': [],
        'last_seen': c.last_seen ? c.last_seen : 0
      }
    })

    .on('data', function (data) {
      ++nvm.rspCnt
      c.last_seen = Math.floor(Date.now() / 1000)
      nvm.socket.setTimeout(0)
      if (checkJSON(data)) {
        nvd = JSON.parse(data)
        ntotal = 0
        ngpu = []
        for (j in nvd.result) {
          ntotal = ntotal + nvd.result[j].speed_sps
          ngpu[j] = {
            'hash': nvd.result[j].speed_sps,
            'hash2': -1,
            'temps': nvd.result[j].temperature,
            'fan': -1,
            'power': nvd.result[j].gpu_power_usage
          }
        }
        nerror = null
        if (ntotal < 1) {
          nerror = '0 HASH'
        }
        ewbf.json[i] = {
          'name': nvm.name,
          'host': hostname(),
          'uptime': nvd.start_time,
          'pools': nvd.current_server,
          'ver': 'EWB',
          'comments': c.comments,
          'offline': c.offline,
          'power': c.power,
          'error': nerror,
          'gpu': ngpu,
          'coin': c.coin,
          'algo': c.algo,
          'hash': ntotal
        }
      } else {
        console.log('bad JSON')
      }
    })

    .on('close', function () {
      setTimeout(poll, nvm.poll)
    })

    .on('error', function (e) {
      ewbf.json[i] = {
        'name': nvm.name,
        'host': hostname(),
        'uptime': '',
        'pools': '',
        'ver': 'EWB',
        'comments': c.comments,
        'offline': c.offline,
        'warning': null,
        'error': e.code,
        'power': '',
        'gpu': [],
        'last_seen': c.last_seen ? c.last_seen : 0
      }
    })

  function poll () {
    nvm.socket.connect(nvm.port, nvm.host)
  };

  if ((typeof c.offline === 'undefined') || !c.offline) {
    poll()
  } else {
    ewbf.json[i] = {
      'name': nvm.name,
      'host': hostname(),
      'uptime': '',
      'pools': '',
      'ver': 'EWB',
      'comments': c.comments,
      'offline': c.offline,
      'error': null,
      'power': '',
      'gpu': []
    }
  }
})

module.exports = {
  getData: function () {
    return ewbf.json
  }
}
// --------------- /REQUESTER ---------------


claymore.js

// --------------- REQUESTER ---------------

var config = require('./config.json')
var net = require('net')

var clay = []
clay.json = []
var lowHashCount = 0

config.clay.forEach(function (item, i, arr) {
  // settings
  var m = clay[i] = {}
  var c = config.clay[i]
  var j = clay.json[i]

  m.name = c.name
  m.host = c.host
  m.port = c.port
  m.poll = (typeof c.poll !== 'undefined') ? c.poll : config.miner_poll
  m.timeout = (typeof c.timeout !== 'undefined') ? c.timeout : config.miner_timeout

  function hostname () {
    return c.hostname ? c.hostname : (m.host + ':' + m.port)
  }

  // stats
  m.reqCnt = 0
  m.rspCnt = 0

  // it was never seen and never found good yet
  c.last_seen = null
  c.last_good = null

  // socket
  m.socket = new net.Socket()

    .on('connect', function () {
      var req = '{"id":0,"jsonrpc":"2.0","method":"miner_getstat1"}'
      ++m.reqCnt
      m.socket.write(req + '\n')
      m.socket.setTimeout(m.timeout)
    })

    .on('timeout', function () {
      m.socket.destroy()
      clay.json[i] = {
        'name': m.name,
        'host': hostname(),
        'uptime': '',
        'pools': '',
        'ver': '',
        'comments': c.comments,
        'offline': c.offline,
        'warning': null,
        'error': 'Error: no response',
        'power': '',
        'gpu': [],
        'last_seen': c.last_seen ? c.last_seen : 0
      }
    })

    .on('data', function (data) {
      ++m.rspCnt
      c.last_seen = Math.floor(Date.now() / 1000)
      m.socket.setTimeout(0)

      var d = JSON.parse(data)
      var eth = d.result[2].split(';')
      var dcr = d.result[4].split(';')
      var eth_hr = d.result[3].split(';')
      var dcr_hr = d.result[5].split(';')
      var temps = d.result[6].split(';')
      var cgpu = []

      for (j = 0; j < eth_hr.length; j++) {
        cgpu[j] = {
          'hash': eth_hr[j],
          'hash2': dcr_hr[j],
          'temps': temps[j * 2],
          'fan': temps[j * 2 + 1],
          'power': -1
        }
      }

      if (eth[0] < 1) {
        lowHashCount++
        cerror = null
        if (lowHashCount > 5) {
          cerror = '0 HASH'
        }
        chash = 0
        chash2 = 0
      } else {
        lowHashCount = 0
        cerror = null
        chash = eth[0]
        chash2 = dcr[0]
      };
      clay.json[i] = {
        'name': m.name,
        'host': hostname(),
        'uptime': Math.floor(Date.now() / 1000) - d.result[1] * 60,
        'pools': d.result[7],
        'ver': d.result[0],
        'comments': c.comments,
        'offline': c.offline,
        'power': c.power,
        'error': cerror,
        'gpu': cgpu,
        'coin': c.coin,
        'algo': c.algo,
        'hash': chash,
        'hash2': chash2
      }
    })

    .on('close', function () {
      setTimeout(poll, m.poll)
    })

    .on('error', function (e) {
      clay.json[i] = {
        'name': m.name,
        'host': hostname(),
        'uptime': '',
        'pools': '',
        'ver': '',
        'comments': c.comments,
        'offline': c.offline,
        'warning': null,
        'error': e.code,
        'power': '',
        'gpu': [],
        'last_seen': c.last_seen ? c.last_seen : 0
      }
    })

  function poll () {
    m.socket.connect(m.port, m.host)
  };

  if ((typeof c.offline === 'undefined') || !c.offline) {
    poll()
  } else {
    clay.json[i] = {
      'name': m.name,
      'host': hostname(),
      'uptime': '',
      'pools': '',
      'ver': '',
      'comments': c.comments,
      'offline': c.offline,
      'error': null,
      'power': '',
      'gpu': []
    }
  }
})

module.exports = {
  getData: function () {
    return clay.json
  }
}
// --------------- /REQUESTER ---------------


server.js
const _ = require('lodash')
const config = require('./config.json')
const TelegramBot = require('node-telegram-bot-api')
const claymore = require('./claymore.js')
const ewbfmonitor = require('./ewbf.js')
const Poloniex = require('./poloniex')
const moment = require('moment')
const bittrex = require('node-bittrex-api')
const https = require('https')

Poloniex.STRICT_SSL = false

const bot = new TelegramBot(config.token, { polling: true })

var ison = []
var oldon = []
var algos = ['Ethash', 'Equihash', 'CryptoNight', 'ClayDual']
var wtmal = ['', '', '']
var suf = ['MH/s', 'H/s', 'H/s', 'MH/s']
var all

var poloniex = new Poloniex(config.poloniex.key, config.poloniex.secret)

bittrex.options({
  'apikey': config.bittrex.key,
  'apisecret': config.bittrex.secret
})

if (config.token !== 0) {
  setInterval(function () {
    var a = ewbfmonitor.getData()
    var b = claymore.getData()
    all = a.concat(b)

    var any = false
    var msgtext = '<pre>'

    if (all.length > 0) {
      for (var miner in all) {
        if (all[miner].error == null) {
          ison[miner] = true
        } else {
          ison[miner] = false
        }
      }

      for (var j in ison) {
        if (ison[j] !== oldon[j]) {
          if (ison[j]) {
            console.log(all[j].name + ' online ' + Date())
            msgtext = msgtext + all[j].name + ' is online\n'
          } else {
            msgtext = msgtext + all[j].name + ' is offline, e:' + all[j].error + '\n'
            console.log(all[j].name + ' offline ' + Date())
          }

          any = true
        }
      }

      oldon = Object.assign({}, ison)
    }

    msgtext = msgtext + '</pre>'

    if (any) {
      bot.sendMessage(config.chatid, msgtext, { parse_mode: 'HTML' })
    }
  }, config.miner_poll)
}

bot.onText(/\/id/, (msg) => {
  bot.sendMessage(msg.chat.id, '<pre>Your chat ID is: ' + msg.chat.id + '</pre>', { parse_mode: 'HTML' })
})

bot.onText(/\/help/, (msg) => {
  var commands = '<b>Commands</b>\n\nMiners\n/status\n/detailed\n/pools\n/ip\n/profit\n\nExchange\n/balances\n/history'

  bot.sendMessage(msg.chat.id, commands, { parse_mode: 'HTML' })
})

bot.onText(/\/ip/, (msg) => {
  var msgtext = '<pre>IP LIST\n'

  for (var f in all) {
    var rig = all[f]
    msgtext = msgtext + rig.name + ': ' + rig.host + ' ' + rig.comments + '\n'
  }

  msgtext = msgtext + '</pre>'

  bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
})

bot.onText(/\/pools/, (msg) => {
  var msgtext = '<pre>POOL LIST\n'

  for (var f in all) {
    var rig = all[f]

    msgtext = msgtext + rig.name + ': ' + rig.pools + ' ' + rig.ver + '\n'
  }

  msgtext = msgtext + '</pre>'

  bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
})

bot.onText(/\/status/, (msg) => {
  var message = '<b>Status</b>\n'
  var rigs = Object.assign({}, all)

  if (_.isEmpty(rigs)) {
    bot.sendMessage(msg.chat.id, 'Mining rigs are not present as yet.')

    return
  }

  for (var f in rigs) {
    var rig = rigs[f]

    message = message + '<b>' + rig.name + '</b>\n'

    if (!_.isEmpty(rig.error)) {
      var lastSeen = moment.unix(rig.last_seen).fromNow()
      message = message + rig.error + ' Last Seen: <i>' + lastSeen + '</i>\n'

      continue
    }

    var primaryHashRate = (rig.hash / 1000).toFixed(2)
    var primaryHashSymbol = (rig.hash > 10000) ? 'MH/s' : 'H/s'
    var secondaryHashRate = (rig.hash2 > 0) ? (rig.hash2 / 1000).toFixed(2) : 0
    var secondaryHashSymbol = (rig.hash2 > 10000) ? 'MH/s' : 'H/s'
    var uptime = moment.unix(rig.uptime).toNow(true)

    message = message + 'Primary Hash Rate: <i>' + primaryHashRate + ' ' + primaryHashSymbol + '</i>\n'

    if (secondaryHashRate > 0) {
      message = message + 'Secondary Hash Rate: <i>' + secondaryHashRate + ' ' + secondaryHashSymbol + '</i>\n'
    }

    message = message + 'Uptime: <i>' + uptime + '</i>\n'
  }

  bot.sendMessage(msg.chat.id, message, { parse_mode: 'HTML' })
})

bot.onText(/\/profit/, (msg) => {
  var rigs = Object.assign({}, all)
  var msgtext = '<pre>PROFIT \n'
  var power = [0, 0, 0, 0]
  var hash = [0, 0, 0, 0]
  var kwprice = config.electricity
  var e = config.currency_symbol
  var wtmpath = 'https://whattomine.com/coins.json?utf8=%E2%9C%93'
  var coinpath = 'https://api.coinmarketcap.com/v1/ticker/bitcoin/?convert=' + config.currency

  for (var f in rigs) {
    var rig = rigs[f]

    if (rig.error === null) {
      if (rig.algo === 'eth') {
        hash[0] = hash[0] + rig.hash * 1
        hash[3] = hash[3] + rig.hash2 * 1
        power[0] = power[0] + rig.power * 1
        wtmal[0] = '&eth=true&factor%5Beth_hr%5D='
      }

      if (rig.algo === 'eq') {
        hash[1] = hash[1] + rig.hash * 1
        power[1] = power[1] + rig.power * 1
        wtmal[1] = '&eq=true&factor%5Beq_hr%5D='
      }

      if (rig.algo === 'cn') {
        hash[2] = hash[2] + rig.hash * 1
        power[2] = power[2] + rig.power * 1
        wtmal[2] = '&cn=true&factor%5Bcn_hr%5D='
      }
    }
  }

  var allpower = 0

  for (var p in power) {
    allpower = allpower + power[p]
  }

  allpower = allpower / 1000
  hash[0] = hash[0] / 1000
  hash[3] = hash[3] / 1000

  for (var a in algos) {
    if (hash[a] > 0) {
      if (a === 0) {
        msgtext = msgtext + algos[a] + ': ' + hash[a].toFixed(2) + suf[a] + ', ' + power[a] + 'W, ' + (hash[a] * 1000 / power[a]).toFixed(2) + 'H/w\n' +
        'Power: d:' + (power[a] * kwprice * 24 / 1000).toFixed(2) + e + ' w:' + (power[a] * kwprice * 24 * 7 / 1000).toFixed(2) + e + ' m:' + (power[a] * kwprice * 24 * 30 / 1000).toFixed(2) + e + '\n'
      }

      if (a === 1 || a === 2) {
        msgtext = msgtext + algos[a] + ': ' + hash[a] + suf[a] + ', ' + power[a] + 'W, ' + (hash[a] / power[a]).toFixed(2) + 'H/w\n' +
        'Power: d:' + (power[a] * kwprice * 24 / 1000).toFixed(2) + e + ' w:' + (power[a] * kwprice * 24 * 7 / 1000).toFixed(2) + e + ' m:' + (power[a] * kwprice * 24 * 30 / 1000).toFixed(2) + e + '\n'
      }
    }
  }

  msgtext = msgtext + 'Total expenses for power: ' + allpower.toFixed(2) + 'kW\n'
  msgtext = msgtext + 'd:' + (allpower * kwprice * 24).toFixed(2) + e + ' w:' + (allpower * kwprice * 24 * 7).toFixed(2) + e + ' m:' + (allpower * kwprice * 24 * 30).toFixed(2) + e + '\n'

  for (var w in wtmal) {
    if (hash[w] > 0) {
      wtmpath = wtmpath + wtmal[w] + hash[w]
    }
  };

  let price = ''
  let profits = ''

  https.get(wtmpath, (res) => {
    res.on('data', (chunk) => {
      profits += chunk
    })

    res.on('end', () => {
      profits = JSON.parse(profits)

      https.get(coinpath, (res) => {
        res.on('data', (chunk) => {
          price += chunk
        })

        res.on('end', () => {
          price = JSON.parse(price)

          var ccurr = 'price_' + config.currency
          var btc = price[0][ccurr]
          var ethtext = 'Ethash:\n'
          var eqtext = 'Equihash:\n'
          var cntext = 'CryptoNight:\n'

          msgtext = msgtext + 'Revenue:\n'

          for (var key in profits['coins']) {
            var profit = profits['coins'][key]
            var pr = profit.tag

            if (pr.length < 6) pr = pr + '    '
            if (pr.length > 5) pr = pr.substr(0, 5)
            if (profit.algorithm === 'Ethash') ethtext = ethtext + pr + ' | d:' + (profit.btc_revenue24 * btc).toFixed(2) + e + ', w:' + (profit.btc_revenue24 * btc * 7).toFixed(0) + e + ', m:' + (profit.btc_revenue24 * btc * 30).toFixed(0) + e + '\n'
            if (profit.algorithm === 'Equihash') eqtext = eqtext + pr + ' | d:' + (profit.btc_revenue24 * btc).toFixed(2) + e + ', w:' + (profit.btc_revenue24 * btc * 7).toFixed(0) + e + ', m:' + (profit.btc_revenue24 * btc * 30).toFixed(0) + e + '\n'
            if (profit.algorithm === 'CryptoNight') cntext = cntext + pr + ' | d:' + (profit.btc_revenue24 * btc).toFixed(2) + e + ', w:' + (profit.btc_revenue24 * btc * 7).toFixed(0) + e + ', m:' + (profit.btc_revenue24 * btc * 30).toFixed(0) + e + '\n'
          }

          var peth = 0
          var peq = 0
          var pecn = 0
          var dayprofit = 0

          if (hash[0] > 0) {
            msgtext = msgtext + ethtext
            peth = profits.coins.Ethereum.btc_revenue24 * 1
          }

          if (hash[1] > 0) {
            msgtext = msgtext + eqtext
            peq = profits.coins.Zcash.btc_revenue24 * 1
          }

          if (hash[2] > 0) {
            msgtext = msgtext + cntext
            pecn = profits.coins.Monero.btc_revenue24 * 1
          }

          dayprofit = (pecn + peq + peth) * btc - allpower * kwprice * 24

          msgtext = msgtext + 'ZEC+ETH+XMR-Electricity:\nd:' + dayprofit.toFixed(2) + e + ', w:' + (dayprofit * 7).toFixed(2) + e + '\nm:' + (dayprofit * 30).toFixed(2) + e + ', y:' + (dayprofit * 365).toFixed(2) + e + '</pre>'

          bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
        })
      }).on('error', (err) => {
        console.log('er: ' + err.message)

        msgtext = '<pre>coinmarketcap.com err: ' + err.message + '</pre>'

        bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
      })
    })
  }).on('error', (err) => {
    console.log('er: ' + err.message)

    msgtext = '<pre>whattomine.com err: ' + err.message + '</pre>'

    bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
  })
})

bot.onText(/\/detailed/, (msg) => {
  let rigs = Object.assign({}, all)
  let message = '<b>Detailed Statistics</b>\n'

  if (_.isEmpty(rigs)) {
    bot.sendMessage(msg.chat.id, 'Mining rigs are not present as yet.')

    return
  }

  for (let f in rigs) {
    let rig = rigs[f]
    let gpuCount = 0

    message += `<b>${rig.name}</b>`

    if (!_.isEmpty(rig.error)) {
      message += ' is offline &#9888;\n'

      continue
    }

    message += ' - ' + (rig.hash / rig.power).toFixed(2) + ' H/W\n'

    for (let i in rig.gpu) {
      let gpu = rig.gpu[i]
      let currency = rig.ver.slice(-3)
      let fanSpeed = (gpu.fan > 0) ? `${gpu.fan}%` : ''
      let gpuPower = (gpu.power > 0) ? `| ${gpu.power} W` : ''
      let hashRate = (currency === 'ETH') ? (gpu.hash / 1000).toFixed(2) + ' MH/s' : `${gpu.hash} H/s`
      let hashPerWatt = (gpu.hash / gpu.power).toFixed(2)

      hashPerWatt = (hashPerWatt > 0) ? ` | ${hashPerWatt} H/W` : ''
      hashRate = (gpu.hash === 0) ? `<b>&#9888; ${hashRate}</b>` : hashRate
      ++gpuCount

      message += ` ${gpuCount} (${currency}): ${hashRate} | ${gpu.temps}&#176;C | ${fanSpeed} ${gpuPower} ${hashPerWatt}\n`
    }
  }

  bot.sendMessage(msg.chat.id, message, { parse_mode: 'HTML' })
})

bot.onText(/\/balances/, (msg) => {
  var e = config.currency_symbol
  var coinpath = 'https://api.coinmarketcap.com/v1/ticker/bitcoin/?convert=' + e
  var price = ''
  var configured = {
    bittrex: config.bittrex.key.length > 0,
    poloniex: config.poloniex.key.length > 0
  }

  if (!configured.bittrex && !configured.poloniex) {
    bot.sendMessage(msg.chat.id, 'No Bittrex or Poloniex accounts configured.')

    return
  }

  https.get(coinpath, (res) => {
    res.on('data', (chunk) => {
      price += chunk
    })

    res.on('end', () => {
      price = JSON.parse(price)

      var ccurr = 'price_' + config.currency
      var btc = price[0][ccurr]
      var msgtext

      if (configured.bittrex) {
        bittrex.getbalances(function (data, err) {
          if (err) {
            console.error(err)

            msgtext = '<pre>bittrex.com err: ' + err.message + '</pre>'

            return
          }

          var bittdata = data.result
          msgtext = '<pre> BITTREX Balances:\n'

          for (var i in bittdata) {
            msgtext = msgtext + bittdata[i].Currency + ': ' + bittdata[i].Balance.toFixed(8) + '\n'
          }

          msgtext = msgtext + '</pre>'

          bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
        })
      }

      if (configured.poloniex) {
        poloniex.returnCompleteBalances(function (err, data) {
          if (err) {
            console.error(err)

            msgtext = '<pre>poloniex.com err: ' + err.message + '</pre>'

            return
          }

          var polodata = data
          var totalvalue = 0
          msgtext = '<pre> POLONIEX Balances:\n'

          for (var key in polodata) {
            var coin = polodata[key].available * 1
            var coinbtc = polodata[key].btcValue * 1
            var name = String(key)

            if (coin > 0) {
              msgtext = msgtext + name + ': ' + coin.toFixed(8) + ', ' + (coinbtc * btc).toFixed(2) + e + '\n'
            }

            totalvalue = totalvalue + coinbtc * 1
          }

          totalvalue = totalvalue * btc
          msgtext = msgtext + 'Total:' + totalvalue.toFixed(2) + e + '</pre>'

          bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
        })
      }
    }).on('error', (err) => {
      console.log('er: ' + err.message)

      var msgtext = '<pre>coinmarketcap.com err: ' + err.message + '</pre>'

      bot.sendMessage(msg.chat.id, msgtext, { parse_mode: 'HTML' })
    })
  })
})

bot.onText(/\/history/, (msg) => {
  var timeNow = Math.floor(Date.now() / 1000)
  var timeLastWeek = timeNow - 604800
  var tLast = new Date(timeLastWeek * 1000).toISOString()
  var msgtext
  var configured = {
    bittrex: config.bittrex.key.length > 0,
    poloniex: config.poloniex.key.length > 0
  }

  if (!configured.bittrex && !configured.poloniex) {
    bot.sendMessage(msg.chat.id, 'No Bittrex or Poloniex accounts configured.')

    return
  }

  bittrex.getwithdrawalhistory({}, function (data, err) {
    if (err) {
      console.error('E: BittrexW' + err)
      return
    }

    var bittw = data.result

    bittrex.getdeposithistory({}, function (data, err) {
      if (err) {
        console.error('E: BittrexD:' + err)

        return
      }

      var bittd = data.result

      msgtext = '<pre> BITTREX WITHDRAWALS:\n'

      for (var i in bittw) {
        var time = moment(bittw[i].Opened).fromNow()

        if (bittw[i].Opened > tLast) {
          msgtext = msgtext + bittw[i].Currency + ': ' + bittw[i].Amount + '\n to:' + bittw[i].Address + '\n date:' + time + '\n'
        }
      }

      msgtext = msgtext + ' BITTREX DEPOSITS:\n'

      for (var deposit in bittd) {
        time = moment(bittd[deposit].LastUpdated).fromNow()

        if (bittd[deposit].LastUpdated > tLast) {
          msgtext = msgtext + bittd[deposit].Currency + ': ' + bittd[deposit].Amount + ', ' + time + '\n'
        }
      }

      bot.sendMessage(msg.chat.id, msgtext + '</pre>', { parse_mode: 'HTML' })
    })
  })

  poloniex.returnDepositsWithdrawals(timeLastWeek, timeNow, function (err, data) {
    if (err) {
      console.log('E: Poloniex:', err)

      return
    }

    var polo = data
    var msgtext = '<pre> POLONIEX WITHDRAWALS:\n'
    var tx
    var time

    for (var i in polo.withdrawals) {
      tx = polo.withdrawals[i]
      time = moment(tx.timestamp * 1000).fromNow()
      msgtext = msgtext + tx.currency + ': ' + tx.amount + '\n to:' + tx.address + '\n date:' + time + '\n'
    }

    msgtext = msgtext + ' POLONIEX DEPOSITS:\n'

    for (var deposit in polo.deposits) {
      tx = polo.deposits[deposit]
      time = moment(tx.timestamp * 1000).fromNow()
      msgtext = msgtext + tx.currency + ': ' + tx.amount + ', ' + time + '\n'
    }

    bot.sendMessage(msg.chat.id, msgtext + '</pre>', { parse_mode: 'HTML' })
  })
})
